### ISSUE TITLE
[Question] How to render a text on rotated pages?

### ISSUE THREAD

**Issue by iSunRise** [gettalong/hexapdf#73]
Hello,

I cannot render horizontal text on rotated page, it becomes vertical instead. Is there any docs how to do it (I also want to render images and lines) or, maybe, I can re-render pdf somehow so pages become really rotated (now there's only rotation flag)?

Thanks,
Alexander

**Comment by gettalong** [comment]
Tl;dr: You need to transform the canvas back to horizontal orientation by using the [`canvas.rotate`](https://hexapdf.gettalong.org/api/HexaPDF/Content/Canvas.html#method-i-rotate) method.

Longer explanation:

This has to do with the drawing model of PDF pages (see section 8 in the PDF 1.7 spec for all the glory details). There is something called the graphics state and it contains, for example, the current line color, the current font, and so on. And it contains the "current transformation matrix" which describes how coordinates are mapped to the page. Note that the `/Rotate` key of the page is also taken into account.

If you want to change a property of the graphics state, you have two options: You can just set the new value, eg. `canvas.stroke_color` but then you need to remember the old value. Or you can save the graphics state with [`canvas.save_graphics_state`](https://hexapdf.gettalong.org/api/HexaPDF/Content/Canvas.html#method-i-save_graphics_state), set the new value and later restore the graphics state.

So, back to the current transformation matrix (CTM). Since this is a matrix, it may be transformed in a way that cannot be reversed. So most PDF writing applications save the graphics state, transform the matrix, apply the operations to the new coordinate system and then restore the graphics state (restoring is actually mandated by the spec but some writers leave this out...).

So if in your case there is currently only the rotation flag set (if I read you correctly), it means that the initial coordinate system is already rotated. So the first thing you need to do is calling `canvas.rotate` to reverse the rotation. Then you can proceed normally.

If rotation is not done via the flag but via page commands, then it would be best to use the [overlay canvas](https://hexapdf.gettalong.org/api/HexaPDF/Type/Page.html#method-i-canvas) to get a blank slate. Note that this will only work correctly if the application that wrote the PDF has restored the graphics state at the end.

**Comment by iSunRise** [comment]
Thanks for the information, I will give it a try!

**Comment by iSunRise** [comment]
In case of interest, I was able to manage the situation using ghostscript. A shell script like:
```
gs -dPrinted -dBATCH -dNOPAUSE -dNOSAFER -q -sDEVICE=pdfwrite output.pdf input.pdf
```
re-generates pdf and sets all `Rotate` flags to 0 but keeps page sizes/orientation as it was "visually" shown.
With re-generated pdf I was able to render horizontal texts on landscape page (it was rendering vertical before due to 90 degree rotation).

Thanks for support anyway!

**Comment by gettalong** [comment]
@iSunRise You can probably do the same with the HexaPDF [page.rotate(0, flatten: true)](https://hexapdf.gettalong.org/api/HexaPDF/Type/Page.html#method-i-rotate) method.
